x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneMinMaxOperator.java

```diff
@@ -0,0 +1,175 @@
+ * Operator that finds the min or max value of a field using Lucene searches
+ * and returns always one entry that mimics the min/max aggregation internal state:
+ * 1. the min/max with a type depending on the {@link NumberType} (The initial value if no doc is seen)
+ * 2. a bool flag (seen) that is true if at least one document has been matched, otherwise false
+ * <p>
+ * It works for fields that index data using lucene {@link PointValues} and/or {@link SortedNumericDocValues}.
+ * It assumes that {@link SortedNumericDocValues} are always present.
+ */
+final class LuceneMinMaxOperator extends LuceneOperator {
+
+    sealed interface NumberType permits LuceneMinFactory.NumberType, LuceneMaxFactory.NumberType {
+
+        /** Extract the competitive value from the {@link PointValues}  */
+        long fromPointValues(PointValues pointValues) throws IOException;
+
+        /** Wraps the provided {@link SortedNumericDocValues} with a {@link MultiValueMode} */
+        NumericDocValues multiValueMode(SortedNumericDocValues sortedNumericDocValues);
+
+        /** Return the competitive value between {@code value1} and {@code value2} */
+        long evaluate(long value1, long value2);
+
+        /** Build the corresponding block */
+        Block buildResult(BlockFactory blockFactory, long result, int pageSize);
+
+        /** Build the corresponding block */
+        Block buildEmptyResult(BlockFactory blockFactory, int pageSize);
+    }
+
+    private static final int PAGE_SIZE = 1;
+
+    private boolean seen = false;
+    private int remainingDocs;
+    private long result;
+
+    private final NumberType numberType;
+
+    private final String fieldName;
+
+    LuceneMinMaxOperator(
+        BlockFactory blockFactory,
+        LuceneSliceQueue sliceQueue,
+        String fieldName,
+        NumberType numberType,
+        int limit,
+        long initialResult
+    ) {
+        super(blockFactory, PAGE_SIZE, sliceQueue);
+        this.remainingDocs = limit;
+        this.numberType = numberType;
+        this.fieldName = fieldName;
+        this.result = initialResult;
+    }
+
+    @Override
+    public boolean isFinished() {
+        return doneCollecting || remainingDocs == 0;
+    }
+
+    @Override
+    public void finish() {
+        doneCollecting = true;
+    }
+
+    @Override
+    public Page getCheckedOutput() throws IOException {
+        if (isFinished()) {
+            assert remainingDocs <= 0 : remainingDocs;
+            return null;
+        }
+        final long start = System.nanoTime();
+        try {
+            final LuceneScorer scorer = getCurrentOrLoadNextScorer();
+            // no scorer means no more docs
+            if (scorer == null) {
+                remainingDocs = 0;
+            } else {
+                final LeafReader reader = scorer.leafReaderContext().reader();
+                final Query query = scorer.weight().getQuery();
+                if (query == null || query instanceof MatchAllDocsQuery) {
```

Comment:
I'm curious if in practice we'll get here often. We'd *prefer* to get the min at local planning time if we've got a `match_all`. But maybe we can't because of a local rewrite or something. That seems pretty plausible.

Reply:
Oh! Deleted documents in only some segments. That's a thing!